File: src/main/java/io/github/kingg22/api/vacunas/panama/service/TokenService.java
package io.github.kingg22.api.vacunas.panama.service;

import io.github.kingg22.api.vacunas.panama.web.dto.PermisoDto;
import io.github.kingg22.api.vacunas.panama.web.dto.UsuarioDto;
import jakarta.validation.constraints.NotNull;
import java.io.Serializable;
import java.time.Duration;
import java.time.Instant;
import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Stream;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
import org.springframework.security.oauth2.jwt.JwsHeader;
import org.springframework.security.oauth2.jwt.JwtClaimsSet;
import org.springframework.security.oauth2.jwt.JwtEncoder;
import org.springframework.security.oauth2.jwt.JwtEncoderParameters;
import org.springframework.stereotype.Service;

/**
 * Service for managing JWT tokens.
 *
 * <p>This service is responsible for encoding information from DTOs entities previously fetched by
 * {@link UsuarioManagementService}, Its sets the issuer and time values based on application.properties.
 *
 * <p>The service now supports both access tokens and refresh tokens, ensuring the correct usage of each token type.
 * Tokens are stored in Redis cache to facilitate efficient validation and retrieval.
 *
 * <p>This service provides methods for validating the existence of tokens in the cache. It does not handler the
 * decoding of tokens; this is managed by the {@link org.springframework.security.oauth2.jwt.JwtDecoder}
 *
 * @see UsuarioManagementService
 * @see org.springframework.security.oauth2.jwt.JwtDecoder
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class TokenService implements ITokenService {
    private final JwtEncoder jwtEncoder;
    private final RedisTemplate<String, Object> redisTemplate;

    @Value("${security.jwt.issuer}")
    private String issuer;

    @Value("${security.jwt.expiration-time}")
    private Integer expirationTime;

    @Value("${security.jwt.refresh-time}")
    private Integer refreshTime;

    public Map<String, Serializable> generateTokens(
            @org.jetbrains.annotations.NotNull UsuarioDto usuarioDto, Map<String, Serializable> idsAdicionales) {
        var data = new LinkedHashMap<String, Serializable>();
        data.put(
                "access_token",
                this.createToken(usuarioDto.id().toString(), getRolesPermisos(usuarioDto), idsAdicionales));
        data.put("refresh_token", this.createRefreshToken(usuarioDto.id().toString()));
        return data;
    }

    public boolean isAccessTokenValid(
            @org.jetbrains.annotations.NotNull @NotNull String userId,
            @org.jetbrains.annotations.NotNull @NotNull String tokenId) {
        var key = generateKey("access", userId, tokenId);
        var hasKeyAccessToken = redisTemplate.hasKey(key);
        if (hasKeyAccessToken == null) {
            throw new IllegalStateException("Redis is unavailable, token validation failed");
        }
        return hasKeyAccessToken;
    }

    public boolean isRefreshTokenValid(
            @org.jetbrains.annotations.NotNull @NotNull String userId,
            @org.jetbrains.annotations.NotNull @NotNull String tokenId) {
        String key = "token:refresh:".concat(userId).concat(":").concat(tokenId);
        Boolean hasKeyRefreshToken = redisTemplate.hasKey(key);
        if (hasKeyRefreshToken == null) {
            throw new IllegalStateException("Redis is unavailable, token validation failed");
        }
        return hasKeyRefreshToken;
    }

    private Collection<String> getRolesPermisos(@org.jetbrains.annotations.NotNull @NotNull UsuarioDto usuarioDto) {
        if (usuarioDto.roles() == null || usuarioDto.roles().isEmpty()) {
            return Collections.emptyList();
        }
        return usuarioDto.roles().stream()
                .flatMap(role -> role != null && role.permisos() != null
                        ? Stream.concat(
                                Stream.of("ROLE_" + role.nombre().toUpperCase()),
                                role.permisos().stream().map(PermisoDto::nombre))
                        : null)
                .toList();
    }

    private String createToken(
            @NotNull String subject, Collection<String> rolesPermisos, Map<String, Serializable> claimsAdicionales) {
        var now = Instant.now();
        var builder = JwtClaimsSet.builder()
                .issuer(issuer)
                .issuedAt(now)
                .notBefore(now)
                .expiresAt(now.plusSeconds(expirationTime))
                .subject(subject)
                .claim("scope", rolesPermisos)
                .id(UUID.randomUUID().toString());

        if (claimsAdicionales != null) {
            for (var claim : claimsAdicionales.entrySet()) {
                if (claim.getValue() != null) {
                    builder.claim(claim.getKey(), claim.getValue());
                }
            }
        }
        var claims = builder.build();
        var header = JwsHeader.with(SignatureAlgorithm.RS256).type("JWT").build();

        var jwtToken = this.jwtEncoder
                .encode(JwtEncoderParameters.from(header, claims))
                .getTokenValue();
        log.debug(
                "created a token for: {}, expires at: {}, id_token: {}",
                subject,
                claims.getExpiresAt(),
                claims.getId());

        saveInCache("access", subject, claims.getId(), jwtToken);
        return jwtToken;
    }

    private String createRefreshToken(@org.jetbrains.annotations.NotNull @NotNull String subject) {
        var now = Instant.now();
        var claims = JwtClaimsSet.builder()
                .issuer(issuer)
                .issuedAt(now)
                .notBefore(now)
                .expiresAt(now.plusSeconds(refreshTime))
                .subject(subject)
                .id(UUID.randomUUID().toString())
                .build();
        var header = JwsHeader.with(SignatureAlgorithm.RS256).type("JWT").build();
        var jwtToken = this.jwtEncoder
                .encode(JwtEncoderParameters.from(header, claims))
                .getTokenValue();
        log.debug(
                "created a refresh token for: {}, expires at: {}, id_token: {}",
                subject,
                claims.getExpiresAt(),
                claims.getId());

        saveInCache("refresh", subject, claims.getId(), jwtToken);
        return jwtToken;
    }

    private void saveInCache(
            @NotNull @org.jetbrains.annotations.NotNull String type,
            @NotNull @org.jetbrains.annotations.NotNull String subject,
            @NotNull @org.jetbrains.annotations.NotNull String id,
            @NotNull @org.jetbrains.annotations.NotNull String jwtToken) {
        redisTemplate.opsForValue().set(generateKey(type, subject, id), jwtToken, Duration.ofSeconds(refreshTime));
    }

    @org.jetbrains.annotations.NotNull
    private String generateKey(
            @NotNull @org.jetbrains.annotations.NotNull String type,
            @NotNull @org.jetbrains.annotations.NotNull String subject,
            @NotNull @org.jetbrains.annotations.NotNull String id) {
        var key = new StringBuilder("token:");
        key.append(type).append(":").append(subject).append(":").append(id);
        return key.toString();
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/usuario/service/TokenService.kt
package io.github.kingg22.api.vacunas.panama.modules.usuario.service

import io.github.kingg22.api.vacunas.panama.modules.usuario.dto.UsuarioDto
import java.io.Serializable

/**
 * Service for managing JWT tokens, including generation and validation.
 *
 * This service is responsible for creating, storing, and validating JWT tokens used in
 * the authentication and authorization processes. It interacts with the [UsuarioService]
 * to fetch user-related data and create access or refresh tokens based on the user's profile and
 * associated data.
 *
 * The generated tokens include both access and refresh tokens, ensuring that the correct type
 * is used for each purpose. The tokens are encoded with data such as user information and expiration
 * times, using values set in the application's configuration (`application.properties`).
 *
 * Tokens are stored in a cache to facilitate fast access and efficient validation during
 * authentication. This ensures that the system can handle large-scale user bases while minimizing
 * redundant calculations.
 */
interface TokenService {
    /**
     * Generates both access and refresh tokens for the given user, using data from the provided DTO.
     * The generated tokens are stored in Redis for fast access and validation.
     *
     * @param usuarioDto DTO containing the user information used to generate the tokens.
     * @param idsAdicionales Additional data (e.g., roles, permissions) to include in the token payload.
     * @return A map containing the generated tokens (access and refresh) as key-value pairs where the key
     *         is the token type (e.g., "access_token", "refresh_token") and the value is the encoded token string.
     */
    suspend fun generateTokens(
        usuarioDto: UsuarioDto,
        idsAdicionales: Map<String, Serializable?> = emptyMap(),
    ): Map<String, Serializable>

    /**
     * Validates the existence of a given access token in Redis cache.
     * This method checks if the provided token ID exists in the cache as a valid access token.
     *
     * @param userId The user ID associated with the token.
     * @param tokenId The unique token ID of the access token to validate.
     * @return A Mono that emits `true` if the token is valid (exists in cache), or `false` otherwise.
     */
    suspend fun isAccessTokenValid(userId: String, tokenId: String): Boolean

    /**
     * Validates the existence of a given refresh token in Redis cache.
     * This method checks if the provided token ID exists in the cache as a valid refresh token.
     *
     * @param userId The user ID associated with the token.
     * @param tokenId The unique token ID of the refresh token to validate.
     * @return A Mono that emits `true` if the token is valid (exists in cache), or `false` otherwise.
     */
    suspend fun isRefreshTokenValid(userId: String, tokenId: String): Boolean
}
